### PR TITLE
CMN-1246 Add group by clause

### PR DESCRIPTION
**Pull request by if-loop69420**
## Summary
What the ticket and the title says

## What's been done?
- Add SyntaxKinds and Tokens for Group By clauses
- Add Rollup/Cube clauses
- Add Grouping-Sets clause
- Add tests for every possible thing
--- 
Me implementing parsing for rollup/cube

### CODE REVIEW

File: src/grammar/query.rs [hunk 1]
```diff
@@ -32,6 +32,10 @@ pub(crate) fn parse_query(p: &mut Parser, expect_into_clause: bool) {
         parse_order_by_clause(p);
     }
 
+    if p.at(T![group]) {
+        parse_group_by_clause(p);
+    }
+
```
Comment (KieranKaelin) [inline src/grammar/query.rs:38]: `GROUP BY` has to appear before `ORDER BY`

File: src/grammar/query.rs [hunk 2]
```diff
@@ -206,6 +210,69 @@ pub(crate) fn parse_order_by_clause(p: &mut Parser) {
...
+    p.expect(T![group]);
+    p.expect(T![by]);
+    safe_loop!(p, {
+        if [T![rollup], T![cube]].contains(&p.current()) {
+            parse_rollup_cube_clause(p);
+        } else if p.at(T![grouping]) {
+            parse_grouping_sets_clause(p);
+        } else {
+            parse_expr(p);
+        }
```
Comment (KieranKaelin) [inline src/grammar/query.rs:224]: Lets do this with a `match` over `p.current()`, should be a bit easier to read :smile:

File: src/grammar/query.rs [hunk 3]
```diff
@@ -206,6 +210,69 @@ pub(crate) fn parse_order_by_clause(p: &mut Parser) {
...
+        parse_expr(p);
+    }
+    p.finish();
+}
+
+pub(crate) fn parse_rollup_cube_clause(p: &mut Parser) {
+    p.start(SyntaxKind::RollupCubeClause);
+    p.expect_one_of(&[T![rollup], T![cube]]);
+    p.expect(T!["("]);
+    // Parse grouping expression list
```
Comment (KieranKaelin) [inline src/grammar/query.rs:240]:
```suggestion
```

File: src/grammar/query.rs [hunk 4]
```diff
@@ -206,6 +210,69 @@ pub(crate) fn parse_order_by_clause(p: &mut Parser) {
...
+            parse_group_expression_list(p);
+        }
+
+        if !p.eat(T![,]) {
+            break;
+        }
+    });
+    p.expect(T![")"]);
+    p.finish();
+}
```
Comment (KieranKaelin) [inline src/grammar/query.rs:263]: Please add a test for `GROUPING SETS`

Comment (KieranKaelin) [reply src/grammar/query.rs:263]: And one for `HAVING` as well

Comment (if-loop69420) [reply src/grammar/query.rs:262]: fn test_group_by_having() {??

**Review by KieranKaelin — APPROVED** [review]

**Pull request merged**